saleor/graphql/csv/mutations.py

```diff
@@ -0,0 +1,92 @@
+import graphene
+from django.core.exceptions import ValidationError
+
+from ...core.permissions import ProductPermissions
+from ...csv import models as csv_models
+from ...csv.utils.export import export_products
+from ..core.enums import CsvErrorCode
+from ..core.mutations import BaseMutation
+from ..core.types.common import CsvError
+from ..product.filters import ProductFilterInput
+from ..product.types import Product
+from ..utils import _resolve_nodes
+from .enums import ExportScope
+from .types import Job
+
+
+class ExportProductsInput(graphene.InputObjectType):
+    scope = ExportScope(
+        description="Determine which products should be exported.", required=True
+    )
+    filter = ProductFilterInput(
+        description="Filtering options for products.", required=False
+    )
+    ids = graphene.List(
+        graphene.NonNull(graphene.ID),
+        description="List of products IDS to export.",
+        required=False,
+    )
+
+
+class ExportProducts(BaseMutation):
+    job = graphene.Field(
+        Job, description="The newly created job which is responsible for export data."
+    )
+
+    class Arguments:
+        input = ExportProductsInput(
+            required=True, description="Fields required to export product data"
+        )
+
+    class Meta:
+        description = "Export products to csv file."
+        permissions = (ProductPermissions.MANAGE_PRODUCTS,)
+        error_type_class = CsvError
+        error_type_field = "csv_errors"
+
+    @classmethod
+    def perform_mutation(cls, root, info, **data):
+        user = info.context.user
+        scope = cls.get_products_scope(info, data["input"])
+        job = csv_models.Job.objects.create(created_by=user)
+        export_products.delay(job.pk, scope)
+        job.refresh_from_db()
+        return cls(job=job)
+
+    @classmethod
+    def get_products_scope(cls, info, input):
+        scope = input["scope"]
+        if scope == ExportScope.IDS.value:
+            return cls.clean_ids(input)
+        elif scope == ExportScope.FILTER.value:
+            return cls.clean_filter(input)
+        return {"all": ""}
+
+    @staticmethod
+    def clean_ids(input):
+        ids = input.get("ids", [])
+        if not ids:
+            raise ValidationError(
+                {
+                    "ids": ValidationError(
+                        "You must provide at least one product id.",
+                        code=CsvErrorCode.REQUIRED.value,
+                    )
+                }
+            )
+        _, pks = _resolve_nodes(ids, graphene_type=Product)
```

Comment:
Maybe we should check if there is a reason why this is private? Dangerous to run alone? Or fine to publicly? If that's fine to run publicly, maybe we should make it public–for now, it is private meaning this wasn't designed to be called outside than the file itself.

Reply:
I don't see the reason why it's private. As far as I can see you are the person who refactored `get_nodes` and created this private method. So it seems you are the right person to give the answer 😄

Reply:
Yes I know I am the author but I don't remember the exact reasons 👀 (git blame strikes)

By the looks of it, I didn't want to give direct access to the raw information, and prevent human-errors while handling the results and raising errors. To prevent:
- Disclosing internal IDs (notice `get_nodes` throws the base64 IDs and not the primary keys, as we don't know what will those contain in the future)
- Mixed error raising/formatting/...
- Prevent mixed logic
- Prevent accidental usage of unwanted complex "utilities"/logics

What is the reason behind you not wanting it to run `get_nodes` as is? It seems fine for your case? Am I missing something?

Reply:
I don't want to use `get_nodes` because it returns queryset and I don't need it there (I just need `pks` of products). It will generate additional useless database query.

Reply:
Okay, I see then. Then, refactor `_resolve_nodes` to `resolve_global_ids_to_primary_keys`